### PR TITLE
Add support for ROWS window frame

### PR DESCRIPTION
**Pull request by pramodsatya**
Add support for:
- ROWS UNBOUNDED PRECEDING
- ROWS CURRENT ROW
- ROWS between UNBOUNDED PRECEDING and UNBOUNDED FOLLOWING
- ROWS between CURRENT ROW and UNBOUNDED FOLLOWING

### CODE REVIEW

**Review by aditi-pandit — COMMENTED** [review]
Please rebase this based on suggestions in the range unbounded preceding/following PR.

**Comment by aditi-pandit** [comment]
Hi Pramod,

  We should work on checking this in next. Please update it to be consistent with your range bounds PR.

**Comment by pramodsatya** [comment]
Hi Aditi, sure, I was waiting for the range unbounded frames PR to be merged so I can rebase this PR and the remaining two PRs for k bounded frames, is that fine?

File: velox/exec/Window.cpp [hunk 1]
```diff
@@ -368,7 +368,10 @@ void Window::callApplyForPartitionRows(
               isStartBound ? rawPeerStarts : rawPeerEnds;
           std::copy(rawPeerBuffer, rawPeerBuffer + numRows, frameBoundBuffer);
         } else {
-          VELOX_NYI("Not supported");
+          std::iota(
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:371]: Please add a comment describing the numbers that are getting populated in this buffer.

**Review by aditi-pandit — COMMENTED** [review]
This code looks good for further review.

File: velox/exec/Window.cpp (hunk 1 above)
Comment (pramodsatya) [reply velox/exec/Window.cpp:373]: Thanks for the suggestion, updated!

File: velox/exec/Window.cpp [hunk 2]
```diff
@@ -368,6 +368,8 @@ void Window::callApplyForPartitionRows(
               isStartBound ? rawPeerStarts : rawPeerEnds;
           std::copy(rawPeerBuffer, rawPeerBuffer + numRows, frameBoundBuffer);
         } else {
+          // Fills the frameBound buffer with increasing values of row indices
```
Comment (aditi-pandit) [inline velox/exec/Window.cpp:371]: Nit -- Reframe "Fills the frameBound buffer with increasing value of row indices (corresponding to CURRENT ROW)  from the startRow of the current output buffer. The startRow has to be adjusted relative to the partition start row."

**Review by aditi-pandit — APPROVED** [review]

File: velox/exec/Window.cpp [hunk 3]
```diff
@@ -368,7 +368,14 @@ void Window::callApplyForPartitionRows(
               isStartBound ? rawPeerStarts : rawPeerEnds;
           std::copy(rawPeerBuffer, rawPeerBuffer + numRows, frameBoundBuffer);
         } else {
-          VELOX_NYI("Not supported");
+          // Fills the frameBound buffer with increasing value of row indices
+          // (corresponding to CURRENT ROW) from the startRow of the current
+          // output buffer. The startRow has to be adjusted relative to the
+          // partition start row.
+          std::iota(
+              frameBoundBuffer,
```
Comment (mbasmanova) [inline velox/exec/Window.cpp:376]: nit: frameBoundBuffer -> rawFrameBounds for consistency

File: velox/functions/prestosql/window/tests/WindowTestBase.h [hunk 4]
```diff
@@ -74,6 +74,16 @@ static std::vector<std::string> kRangeFrameClauses = {
     "range between unbounded preceding and unbounded following",
 };
 
+/// Common set of window function frame clauses in ROWS mode, with current row,
+/// unbounded preceding, and unbounded following frame combinations.
+static std::vector<std::string> kRowFrameClauses = {
```
Comment (mbasmanova) [inline velox/functions/prestosql/window/tests/WindowTestBase.h:79]: Is this same as kRangeFrameClauses except for s/range/rows?

kRowFrameClauses -> kRowsFrameClauses (add 's' to Rows)

**Review by mbasmanova — APPROVED** [review]
@pramodsatya Looks good to me % minor comments below.

File: velox/functions/prestosql/window/tests/WindowTestBase.h [hunk 5]
```diff
@@ -74,6 +74,16 @@ static std::vector<std::string> kRangeFrameClauses = {
...
+    "rows unbounded preceding",
+    "rows current row",
+    "rows between unbounded preceding and current row",
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/functions/prestosql/window/tests/WindowTestBase.h:82]: Same for kRangeFrameClauses

[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/WindowTestBase.h:82]: Yes they are both the same but the intention was to test all possible frame combinations so both of them were added. Should I remove "rows between unbounded preceding and current row" frame clause?

File: velox/functions/prestosql/window/tests/WindowTestBase.h (hunk 4 above)
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/WindowTestBase.h]: Yes, they are identical except for range/rows but in subsequent PRs, tests with k preceding and k following frame bounds will be added to kRowsFrameClauses so separate variables are used.

File: velox/functions/prestosql/window/tests/WindowTestBase.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/window/tests/WindowTestBase.h]: Done, thanks for the suggestion!

**Review by mbasmanova — COMMENTED** [review]
Thanks.

**Comment by aditi-pandit** [comment]
https://github.com/facebookincubator/velox/issues/685

**Pull request closed without merging**